sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move

```diff
@@ -0,0 +1,225 @@
+    use Sui::Balance::Balance;
+    use Sui::TxContext::{Self, TxContext};
+    use Stable::RegulatedCoin::{Self as C, RegulatedCoin};
+
+    struct FREE has drop {}
+
+    // === implement the interface for the RegulatedCoin ===
+
+    public fun borrow(coin: &RegulatedCoin<FREE>): &Balance<FREE> { C::borrow(FREE {}, coin) }
+    public fun borrow_mut(coin: &mut RegulatedCoin<FREE>): &mut Balance<FREE> { C::borrow_mut(FREE {}, coin) }
+    public fun from_balance(balance: Balance<FREE>, ctx: &mut TxContext): RegulatedCoin<FREE> { C::from_balance(FREE {}, balance, TxContext::sender(ctx), ctx) }
+    public fun into_balance(coin: RegulatedCoin<FREE>): Balance<FREE> { C::into_balance(FREE {}, coin) }
+
+    // === and that's it (+ minting and currency creation) ===
+}
+
+// A very RESTricted coin.
+module Stable::REST {
+    use Stable::RegulatedCoin::{Self as C, RegulatedCoin};
+
+    use Sui::Balance::{Self, Balance};
+    use Sui::Coin::{Self, TreasuryCap};
+    use Sui::TxContext::{Self, TxContext};
+    use Sui::Transfer;
+
+    const ENotImplemented: u64 = 0;
+    const ENotAllowed: u64 = 1;
+    const ENotOwner: u64 = 2;
+
+    struct REST has drop {}
+
+    /// A restricted transfer of the Balance
+    struct CoinTransfer has key {
+        id: Sui::ID::VersionedID,
+        balance: Balance<REST>,
+        to: address
+    }
+
+    /// Currently let's just use Coin::TreasuryCap functionality
+    fun init(ctx: &mut TxContext) {
+        Transfer::transfer(
+            Coin::create_currency(REST {}, ctx),
+            TxContext::sender(ctx)
+        )
+    }
+
+    /// Only owner of the treasury cap can create new Balances; for example, after a KYC process;
+    public fun create_empty_for(_cap: &TreasuryCap<REST>, for: address, ctx: &mut TxContext) {
+        Transfer::transfer(C::zero(REST {}, for, ctx), for)
+    }
+
+    /// Allow borrowing as is, by default
+    public fun borrow(coin: &RegulatedCoin<REST>): &Balance<REST> { C::borrow(REST {}, coin) }
+    public fun borrow_mut(coin: &mut RegulatedCoin<REST>, ctx: &mut TxContext): &mut Balance<REST> {
+        assert!(TxContext::sender(ctx) == C::owner(coin), ENotOwner); // only owner can access the balance
+        C::borrow_mut(REST {}, coin)
+    }
+
+    // === Coin Transfers ===
+
+    public(script) fun transfer(
+        coin: &mut RegulatedCoin<REST>, value: u64, to: address, ctx: &mut TxContext
+    ) {
+        Transfer::transfer(CoinTransfer {
+            id: TxContext::new_id(ctx),
+            balance: Balance::split(borrow_mut(coin, ctx), value),
+            to
+        }, to)
+    }
+
+    public(script) fun accept_transfer(
+        coin: &mut RegulatedCoin<REST>, transfer: CoinTransfer, ctx: &mut TxContext
+    ) {
+        let CoinTransfer { id, balance, to } = transfer;
+        assert!(C::owner(coin) == to, ENotOwner);
+        Balance::join(borrow_mut(coin, ctx), balance);
+        Sui::ID::delete(id);
+    }
+
+    // === Explicit "Not Implemented" part ===
```

Comment:
Wondering if we need/want this. For a client, I assume it would be easiest find out some functionality isn't supported by a module by checking whether it has the function(s) in question. Seeing that the functions exist, but discovering that they abort when trying them out seems less nice IMO.

Reply:
Yeah! I just meant it as an interface-like thing, and left it just to explicitly show that this is intentional. I wouldn’t pay additional gas for that if I were publishing it on the chain.